File: doc-examples/azure-cosmos-example-java/src/main/java/example/FamilyRepository.java
package example;

import com.azure.cosmos.models.PartitionKey;
import io.micronaut.context.annotation.Parameter;
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.data.annotation.Id;
import io.micronaut.data.annotation.Join;
import io.micronaut.data.annotation.Query;
import io.micronaut.data.cosmos.annotation.CosmosRepository;
import io.micronaut.data.model.jpa.criteria.PersistentEntityCriteriaBuilder;
import io.micronaut.data.repository.PageableRepository;
import io.micronaut.data.repository.jpa.JpaSpecificationExecutor;
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification;

import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@CosmosRepository
public abstract class FamilyRepository implements PageableRepository<Family, String>, JpaSpecificationExecutor<Family> {

    @NonNull
    public abstract Optional<Family> findById(String id);

    public abstract void updateRegistered(@Id String id, boolean registered);

    public abstract void updateAddress(@Parameter("id") @Id String id, @NonNull @Parameter("address") Address address);

    // Raw query for Cosmos update is not supported and calling this method will throw an error.
    @Query("UPDATE family f SET f.lastName=@p1 WHERE f.id=@p2")
    public abstract void updateLastName(String id, String lastName);

    // tag::partitionkey[]
    public abstract Optional<Family> queryById(String id, PartitionKey partitionKey);

    public abstract void updateRegistered(@Id String id, boolean registered, PartitionKey partitionKey);

    public abstract void deleteByLastName(String lastName, PartitionKey partitionKey);

    public abstract void deleteById(String id, PartitionKey partitionKey);
    // end::partitionkey[]

    // Raw query not supported for delete so this would throw an error
    @Query("DELETE FROM family f WHERE f.registered=@p1")
    public abstract void deleteByRegistered(boolean registered);

    public abstract boolean existsByIdAndRegistered(String id, boolean registered);

    public abstract long countByRegistered(boolean registered);

    @Query("SELECT VALUE f.lastName FROM family f ORDER BY f.lastName DESC OFFSET 0 LIMIT 1")
    public abstract String lastOrderedLastName();

    @Query("SELECT VALUE f.registeredDate FROM family f WHERE NOT IS_NULL(f.registeredDate) ORDER BY f.registeredDate DESC OFFSET 0 LIMIT 1")
    public abstract Date lastOrderedRegisteredDate();

    // tag::relations[]
    public abstract List<Family> findByAddressStateAndAddressCityOrderByAddressCity(String state, String city);

    public abstract void updateByAddressCounty(String county, boolean registered, @Nullable Date registeredDate);

    @Join(value = "children.pets", alias = "pets")
    public abstract List<Family> findByChildrenPetsType(PetType type);

    public abstract List<Child> findChildrenByChildrenPetsGivenName(String name);
    // end::relations[]

    public abstract List<Family> findByIdIn(List<String> ids);

    public abstract List<Family> findByIdNotIn(List<String> ids);

    public abstract List<Family> findByLastNameLike(String lastName);

    // tag::method_array_contains[]
    public abstract List<Family> findByTagsArrayContains(String tag);
    // end::method_array_contains[]

    // tag::array_contains_partial[]
    @Query("SELECT DISTINCT VALUE f FROM family f WHERE ARRAY_CONTAINS(f.children, :gender, true)")
    public abstract List<Family> childrenArrayContainsGender(Map.Entry<String, Object> gender);
    // end::array_contains_partial[]

    static class Specifications {

        public static PredicateSpecification<Family> lastNameEquals(String lastName) {
            return (root, criteriaBuilder) -> criteriaBuilder.equal(root.get("lastName"), lastName);
        }

        public static PredicateSpecification<Family> idsIn(String... ids) {
            return (root, criteriaBuilder) -> root.get("id").in(Arrays.asList(ids));
        }

        public static PredicateSpecification<Family> idsNotIn(String... ids) {
            return (root, criteriaBuilder) -> root.get("id").in(Arrays.asList(ids)).not();
        }

        public static PredicateSpecification<Family> idsInAndNotIn(List<String> idsIn, List<String> idsNotIn) {
            return (root, criteriaBuilder) -> criteriaBuilder.and(root.get("id").in(idsIn), root.get("id").in(idsNotIn).not());
        }

        // tag::predicate_array_contains[]
        public static PredicateSpecification<Family> tagsContain(String tag) {
            return (root, criteriaBuilder) -> ((PersistentEntityCriteriaBuilder) criteriaBuilder).arrayContains(root.get("tags"), criteriaBuilder.literal(tag));
        }
        // end::predicate_array_contains[]

        // tag::predicate_array_contains_partial[]
        public static PredicateSpecification<Family> childrenArrayContainsGender(GenderAware gender) {
            return (root, criteriaBuilder) -> ((PersistentEntityCriteriaBuilder) criteriaBuilder).arrayContains(root.join("children"), criteriaBuilder.literal(gender));
        }
        // end::predicate_array_contains_partial[]
    }
}

File: doc-examples/azure-cosmos-example-kotlin/src/main/kotlin/example/FamilyRepository.kt
package example

import com.azure.cosmos.models.PartitionKey
import io.micronaut.context.annotation.Parameter
import io.micronaut.core.annotation.NonNull
import io.micronaut.core.annotation.Nullable
import io.micronaut.data.annotation.Id
import io.micronaut.data.annotation.Join
import io.micronaut.data.annotation.Query
import io.micronaut.data.cosmos.annotation.CosmosRepository
import io.micronaut.data.model.jpa.criteria.PersistentEntityCriteriaBuilder
import io.micronaut.data.repository.PageableRepository
import io.micronaut.data.repository.jpa.JpaSpecificationExecutor
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification
import jakarta.persistence.criteria.CriteriaBuilder
import jakarta.persistence.criteria.Root
import java.util.*

@CosmosRepository
abstract class FamilyRepository : PageableRepository<Family, String>, JpaSpecificationExecutor<Family> {
    @NonNull
    abstract override fun findById(id: String): Optional<Family>
    abstract fun updateRegistered(@Id id: String, registered: Boolean)
    abstract fun updateAddress(@Parameter("id") @Id id: String, @NonNull @Parameter("address") address: Address)

    // Raw query for Cosmos update is not supported and calling this method will throw an error.
    @Query("UPDATE family f SET f.lastName=@p1 WHERE f.id=@p2")
    abstract fun updateLastName(id: String, lastName: String)

    // tag::partitionkey[]
    abstract fun queryById(id: String?, partitionKey: PartitionKey?): Optional<Family?>?
    abstract fun deleteByLastName(lastName: String, partitionKey: PartitionKey)
    abstract fun deleteById(id: String, partitionKey: PartitionKey)
    abstract fun updateRegistered(@Id id: String, registered: Boolean, partitionKey: PartitionKey)
    // end::partitionkey[]

    // Raw query not supported for delete so this would throw an error
    @Query("DELETE FROM family f WHERE f.registered=@p1")
    abstract fun deleteByRegistered(registered: Boolean)
    abstract fun existsByIdAndRegistered(id: String, registered: Boolean): Boolean
    abstract fun countByRegistered(registered: Boolean): Long
    @Query("SELECT VALUE f.lastName FROM family f ORDER BY f.lastName DESC OFFSET 0 LIMIT 1")
    abstract fun lastOrderedLastName(): String?
    @Query("SELECT VALUE f.registeredDate FROM family f WHERE NOT IS_NULL(f.registeredDate) ORDER BY f.registeredDate DESC OFFSET 0 LIMIT 1")
    abstract fun lastOrderedRegisteredDate(): Date?
    // tag::relations[]
    abstract fun findByAddressStateAndAddressCityOrderByAddressCity(state: String, city: String): List<Family>
    abstract fun updateByAddressCounty(county: String, registered: Boolean, @Nullable registeredDate: Date?)
    @Join(value = "children.pets", alias = "pets")
    abstract fun findByChildrenPetsType(type: PetType): List<Family>
    abstract fun findChildrenByChildrenPetsGivenName(name: String): List<Child>
    // end::relations[]
    abstract fun findByIdIn(ids: List<String>): List<Family>
    abstract fun findByIdNotIn(ids: List<String>): List<Family>
    abstract fun findByLastNameLike(lastName: String): List<Family>

    // tag::method_array_contains[]
    abstract fun findByTagsArrayContains(tag: String): List<Family>
    // end::method_array_contains[]

    // tag::array_contains_partial[]
    @Query("SELECT DISTINCT VALUE f FROM family f WHERE ARRAY_CONTAINS(f.children, :gender, true)")
    abstract fun childrenArrayContainsGender(gender: Map.Entry<String, Any>): List<Family>
    // end::array_contains_partial[]

    internal object Specifications {
        fun lastNameEquals(lastName: String): PredicateSpecification<Family> {
            return PredicateSpecification { root: Root<Family>, criteriaBuilder: CriteriaBuilder ->
                criteriaBuilder.equal(
                    root.get<Any>("lastName"),
                    lastName
                )
            }
        }

        fun idsIn(vararg ids: String): PredicateSpecification<Family> {
            return PredicateSpecification { root: Root<Family>, _: CriteriaBuilder ->
                root.get<Any>("id").`in`(
                    listOf(*ids)
                )
            }
        }

        fun idsNotIn(vararg ids: String): PredicateSpecification<Family> {
            return PredicateSpecification { root: Root<Family>, _: CriteriaBuilder ->
                root.get<Any>("id").`in`(
                    listOf(*ids)
                ).not()
            }
        }

        fun idsInAndNotIn(idsIn: List<String>, idsNotIn: List<String>): PredicateSpecification<Family> {
            return PredicateSpecification { root: Root<Family>, criteriaBuilder: CriteriaBuilder ->
                criteriaBuilder.and(
                    root.get<Any>("id").`in`(idsIn),
                    root.get<Any>("id").`in`(idsNotIn).not()
                )
            }
        }

        // tag::predicate_array_contains[]
        fun tagsContain(tag: String): PredicateSpecification<Family?>? {
            return PredicateSpecification { root: Root<Family?>, criteriaBuilder: CriteriaBuilder ->
                (criteriaBuilder as PersistentEntityCriteriaBuilder).arrayContains(
                    root.get<Any>("tags"),
                    criteriaBuilder.literal(tag)
                )
            }
        }
        // end::predicate_array_contains[]

        // tag::predicate_array_contains_partial[]
        fun childrenArrayContainsGender(gender: IGenderAware): PredicateSpecification<Family?>? {
            return PredicateSpecification { root: Root<Family?>, criteriaBuilder: CriteriaBuilder ->
                (criteriaBuilder as PersistentEntityCriteriaBuilder).arrayContains(
                    root.join<Any, Any>("children"),
                    criteriaBuilder.literal(gender)
                )
            }
        }
        // end::predicate_array_contains_partial[]
    }
}
